_specifications/lsp/3.18/language/completion.md

```diff
@@ -166,6 +166,18 @@ export interface CompletionClientCapabilities {
 		 * @since 3.17.0
 		 */
 		itemDefaults?: string[];
+
+		/**
+		 * Specifies which fields of `CompletionList.applyKind` the client
+		 * supports. If omitted, no properties are supported and all fields
+		 * in a completion item will replace the defaults.
+		 * 
+		 * Clients may only specify fields that have merge rules defined in the
+		 * LSP spec.
+		 *
+		 * @since 3.18.0
+		 */
+		applyKinds?: string[];
```

Comment:
This seems to rely on `itemDefaults` capability being supported. Would it make it more obvious if it would be called something like `itemDefaultsApplyKinds`.

And another discussion could be had about the "applyKinds" naming. It doesn't feel very descriptive to me (perhaps subjective). Also it refers to "kind" which is a concept that is already present in completion items that is not related to "kinds" here. Perhaps a better one would be `itemDefaultsMergeStrategy" (although it might be weird to call it "merge strategy" when the strategy can either replace or merge.

Reply:
I don't feel strongly about the name, though I think like `apply` more than `merge` since as you say, merge is one option. Maybe something like "combine" is a bit clearer? I do agree `kind` maybe could be clearer (strategy? method?)

I had a scan through the spec to see if there's anything similar we could name it the same as, but I couldn't find anything that feels the same.

Reply:
I would name it `applyKindSupport: boolean`. We should also add a `ApplyKind` definition since the spec usually doesn't use inline string or types something like

```
export namespace ApplyKind {
        export const Replace: 'replace' = 'replace;
	export const Merge: 'merge' = 'merge';
}

export type ApplyKind = ApplyKind.Replace | ApplyKind.Merge;
```

I do think one boolean is enough due to: if a client supports the new filed (e.g. via itemDefaults) and it opts into `applyKindSupport`  it MUST then support the new filed in the applyKind literal. I think that is fair.

Reply:
And you are fine with using the `kind` wording even though completion items already have a [concept of `kind`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionItemKind) that is something completely different?

Reply:
@dbaeumer I've pushed changes to this affect - PTAL! I did wonder if we should use `ApplyKind` or `Completion(Item?)ApplyKind`.. I don't know if we might support replace/merge elsewhere in future and if so we'd want to use the same enum or not?

@rchl 

> And you are fine with using the kind wording even though completion items already have a [concept of kind](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionItemKind)

There are a lot of `kind`s in the spec so I don't think reusing kind is bad (it's `CompletionItemKind` vs `ApplyKind`), although I'll note we also have `InsertTextMode` which is a "mode", so that's possibly another option?